flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -881,12 +859,16 @@ private boolean areCheckpointsWithFinishedTasksEnabled() {
     }
 
     @Override
-    public void cleanUp() throws Exception {
+    public final void cleanUp() throws Exception {
         LOG.debug(
                 "Cleanup StreamTask (operators closed: {}, cancelled: {})",
                 closedOperators,
                 canceled);
 
+        failing = !canceled;
+
+        Exception suppressedException = runAndSuppressThrowable(this::cancelTask, null);
```

Comment:
Do you sure that it is ok to cancelTask in cleanUp even if the canceled flag doesn't set?

Reply:
In the current (master) version, the equivalent call from `runWithCleanUpOnFail` is **only** made if the task is **not** cancelled. That is, we call `cancelTask` if it wasn't cancelled **yet**.

(regardless of which thread called `cancelTask()`, cleanup waits for `StreamTask.completionFuture` to complete) 

So the check here should actually be: 
```
if (!cancelled) {
        Exception suppressedException = runAndSuppressThrowable(this::cancelTask, null);
}
```

However, it doesn't prevent double cancellation. I'm going to change `boolean canceled` from `volatile` to `AtomicBoolean` (and use CAS when calling `cancelTask()`). WDYT?

Reply:
The current master is call `cancelTask` if the **exception happens** and the task is **not canceled**. But in your new version you do it always regardless the exception or not. So as I understand it is not equvialent call right now it is new behaviour. This why I have concern about the case when the task sucessfully finished but we anyway call the `cancelTask`. Right now I don't understand why we need it.

Reply:
Yes, you are right.
I've changed added `Throwable` argument to `cleanUp`, and only `cancelTask` if it's not null.